cpp/src/arrow/filesystem/azurefs.h

```diff
@@ -0,0 +1,161 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <azure/core/credentials/credentials.hpp>
+#include <azure/storage/common/storage_credential.hpp>
```

Comment:
Hmm, we definitely don't want to include such third-party headers in our headers.
Instead this should be hidden in the `.cc` file (see `s3fs.h` for example).

Reply:
` std::shared_ptr<Azure::Storage::StorageSharedKeyCredential>
      storage_credentials_provider;`
` std::shared_ptr<Azure::Core::Credentials::TokenCredential>
      service_principle_credentials_provider;`
These two member variables of AzureOptions class use these header files.

Reply:
These can be forward-declared just like the equivalent classes in `s3fs.h`.

Reply:
Or we can hide them behind a private implementation class with a pimpl-like idiom.